src/librustc/traits/error_reporting/suggestions.rs

```diff
@@ -583,122 +601,120 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
 
             let tables = self.in_progress_tables.map(|t| t.borrow()).unwrap();
 
-            if let hir::FunctionRetTy::Return(ret_ty) = sig.decl.output {
-                let mut all_returns_conform_to_trait = true;
-                let mut all_returns_have_same_type = true;
-                let mut last_ty = None;
-                if let Some(ty_ret_ty) = tables.node_type_opt(ret_ty.hir_id) {
-                    let cause = ObligationCause::misc(ret_ty.span, ret_ty.hir_id);
-                    if let ty::Dynamic(predicates, _) = &ty_ret_ty.kind {
-                        for predicate in predicates.iter() {
-                            for expr in &visitor.0 {
-                                if let Some(returned_ty) = tables.node_type_opt(expr.hir_id) {
-                                    if let Some(ty) = last_ty {
-                                        all_returns_have_same_type &= ty == returned_ty;
-                                    }
-                                    last_ty = Some(returned_ty);
-
-                                    let param_env = ty::ParamEnv::empty();
-                                    let pred = predicate.with_self_ty(self.tcx, returned_ty);
-                                    let obligation =
-                                        Obligation::new(cause.clone(), param_env, pred);
-                                    all_returns_conform_to_trait &=
-                                        self.predicate_may_hold(&obligation);
-                                }
-                            }
-                        }
-                    }
-                } else {
-                    // We still want to verify whether all the return types conform to each other.
+            let mut all_returns_conform_to_trait = true;
+            let mut all_returns_have_same_type = true;
+            let mut last_ty = None;
+            if let Some(ty_ret_ty) = tables.node_type_opt(ret_ty.hir_id) {
+                let cause = ObligationCause::misc(ret_ty.span, ret_ty.hir_id);
+                let param_env = ty::ParamEnv::empty();
+                if let ty::Dynamic(predicates, _) = &ty_ret_ty.kind {
                     for expr in &visitor.0 {
                         if let Some(returned_ty) = tables.node_type_opt(expr.hir_id) {
-                            if let Some(ty) = last_ty {
-                                all_returns_have_same_type &= ty == returned_ty;
-                            }
+                            all_returns_have_same_type &=
+                                Some(returned_ty) == last_ty || last_ty.is_none();
```

Comment:
Hmm; I guess it was cleaner before with the `if let` if you have to include the `||` which it does seem like you do. Sorry :)